andela/codepirates-ah-backend
# Ft api social login 167313401

#### What does this PR do?

  Add the social login feature

#### Description of Task to be completed?

  - Login from social account if user exists
  - Create user and login if social user does not exist

#### How should this be manually tested?

  - If in a local environment, clone and install this repo locally and start the server with nom start/run dev common
  - In a browser enter the url http://:base-url/login/:social-account
    - The base-url is localhost:3000 for local testing and https://codepirates-ah-backend.herokuapp.com/ for remote testing
    - Social-account is facebook or google or twitter for the respective logins
  - In the resulting window; [facebook](https://user-images.githubusercontent.com/36167006/62922655-3869e780-bdac-11e9-8931-ee307bc5c380.png), [google](https://user-images.githubusercontent.com/36167006/62927228-6522fc80-bdb6-11e9-91a1-b5d6ff52d046.png), [twitter](https://user-images.githubusercontent.com/36167006/62923881-102fb800-bdaf-11e9-81c4-ef976d9d132d.png), enter your social media account credentials
  - If the account does not exist on author's haven, you will receive a [response](https://user-images.githubusercontent.com/36167006/62924005-74eb1280-bdaf-11e9-9bb4-f6491b6812eb.png) to consent to creating an account with Author's Haven
    - Enter localhost:3000/signup/social in the address bar and hit enter, you will see the successful signup message in the [response page](https://user-images.githubusercontent.com/36167006/62924131-b7acea80-bdaf-11e9-98e2-9a1b8b593412.png)
    - Open your associated email and confirm the address
    - Continue with theses steps
  - If the user account already exists on Author's Haven, you will  receive the [login success](https://user-images.githubusercontent.com/36167006/62922821-a57d7d00-bdac-11e9-891a-77a3678b9877.png) response containing your particulars and a token


#### Any background context you want to provide?

#### What are the relevant pivotal tracker stories?

  [finishes #167313401](https://www.pivotaltracker.com/story/show/167313401)

#### Screenshots (if appropriate)


facebook login page after entering localhost:3000/login/facebook in browser address bar in incognito window

<img width="1061" alt="facebook login page on virgin browser" src="https://user-images.githubusercontent.com/36167006/62922655-3869e780-bdac-11e9-8931-ee307bc5c380.png">


Alternative on personalised browser

<img width="1103" alt="possible facebook interface on given browser" src="https://user-images.githubusercontent.com/36167006/62922702-53d4f280-bdac-11e9-8bd8-3f8a24688951.png">


google login window on incognito window after entering localhost:3000/login/google in browser address bar

<img width="1396" alt="Screen Shot 2019-08-13 at 10 36 27" src="https://user-images.githubusercontent.com/36167006/62927228-6522fc80-bdb6-11e9-91a1-b5d6ff52d046.png">


Google login on personalised browser after entering localhost:3000/login/google in browser address bar

<img width="1125" alt="google login" src="https://user-images.githubusercontent.com/36167006/62922765-79fa9280-bdac-11e9-8208-2ef0474548bc.png">


Twitter login window after entering localhost:3000/login/twitter in browser address bar

<img width="936" alt="Screen Shot 2019-08-13 at 09 43 55" src="https://user-images.githubusercontent.com/36167006/62923881-102fb800-bdaf-11e9-81c4-ef976d9d132d.png">


Response prompting non existing social user consent for signup

<img width="1423" alt="Screen Shot 2019-08-13 at 09 47 30" src="https://user-images.githubusercontent.com/36167006/62924005-74eb1280-bdaf-11e9-9bb4-f6491b6812eb.png">


Response window on successful signup of social user

<img width="1425" alt="Screen Shot 2019-08-13 at 09 49 37" src="https://user-images.githubusercontent.com/36167006/62924131-b7acea80-bdaf-11e9-98e2-9a1b8b593412.png">


successful login response

<img width="1178" alt="Successful social login response" src="https://user-images.githubusercontent.com/36167006/62922821-a57d7d00-bdac-11e9-891a-77a3678b9877.png">

#### Questions:
